core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -370,13 +370,32 @@ void CudaInternal::fence() const {
   fence("Kokkos::CudaInternal::fence(): Unnamed Instance Fence");
 }
 
-void CudaInternal::initialize(cudaStream_t stream, bool manage_stream) {
+void CudaInternal::initialize(int cuda_device, cudaStream_t stream,
+                              bool manage_stream) {
+  // Check that the device associated with the stream matches cuda_device
+  CUcontext context;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuStreamGetCtx(stream, &context)));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuCtxPushCurrent(context)));
+  int device_for_stream;
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuCtxGetDevice(&device_for_stream)));
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaError_t(cuCtxPopCurrent(&context)));
+
+  if (device_for_stream != cuda_device) {
+    std::stringstream ss;
+    ss << "Error: The provided stream is associated with device "
+       << device_for_stream << " but device " << cuda_device
+       << " was requested in the execution space instance constructor!";
+    Kokkos::abort(ss.str().c_str());
+  }
```

Comment:
So if I'm a user and I want to use multiple devices, how am I meant to do this? If I need to set a device and then create a stream on that device and then call the new `Cuda(dev_id,stream)` to create the space, we can get all that information from just one of these arguments, so would it make since to just take a stream and query the device_id internally?